README.md
rackspace_rsyslog Cookbook
================

Installs and configures rsyslog to replace sysklogd for client and/or server use. By default, the service will be configured to log to files on local disk. See the Recipes and Examples sections for other uses.


Requirements
------------
### Platforms
Tested on:
- Ubuntu 12.04
- RedHat 6.4
- Debian 7.2

### Other
To use the `recipe[rackspace_rsyslog::client]` recipe, you'll need to set up the `['rackspace_rsyslog']['server_search']` or `['rackspace_rsyslog']['server_ip']` attributes.  See the __Recipes__ and __Examples__ sections below.

Attributes
----------
See `attributes/default.rb` for default values.

* `node['rackspace_rsyslog']['config']['log_dir']` - If the node is an rsyslog server, this specifies the directory where the logs should be stored.
* `node['rackspace_rsyslog']['config']['server']` - Determined automaticaly and set to true on the server.
* `node['rackspace_rsyslog']['config']['server_ip']` - If not defined then search will be used to determine rsyslog server. Default is `nil`.  This can be a string or an array.
* `node['rackspace_rsyslog']['config']['server_search']` - Specify the criteria for the server search operation. Default is `role:loghost`.
* `node['rackspace_rsyslog']['config']['protocol']` - Specify whether to use `udp` or `tcp` for remote loghost. Default is `tcp`.
* `node['rackspace_rsyslog']['config']['port']` - Specify the port which rsyslog should connect to a remote loghost.
* `node['rackspace_rsyslog']['config']['remote_logs']` - Specify wether to send all logs to a remote server (client option). Default is `true`.
* `node['rackspace_rsyslog']['config']['per_host_dir']` - "PerHost" directories for template statements in `35-server-per-host.conf`. Default value is the previous cookbook version's value, to preserve compatibility. See __server__ recipe below.
* `node['rackspace_rsyslog']['config']['priv_seperation']` - Whether to use privilege seperation or not.
* `node['rackspace_rsyslog']['config']['max_message_size']` - Specify the maximum allowed message size. Default is 2k.
* `node['rackspace_rsyslog']['config']['user']` - Who should own the configuration files and directories
* `node['rackspace_rsyslog']['config']['group']` - Who should group-own the configuration files and directories
* `node['rackspace_rsyslog']['config']['defaults_file']` - The full path to the defaults/sysconfig file for the service.
* `node['rackspace_rsyslog']['config']['service_name']` - The platform-specific name of the service
* `node['rackspace_rsyslog']['config']['preserve_fqdn']` - Value of the `$PreserveFQDN` configuration directive in `/etc/rsyslog.conf`. Default is 'off' for compatibility purposes.
* `node['rackspace_rsyslog']['config']['high_precision_timestamps']` -  Enable high precision timestamps, instead of the "old style" format.  Default is 'false'.
* `node['rackspace_rsyslog']['config']['repeated_msg_reduction']` -  Value of `$RepeatedMsgReduction` configuration directive in `/etc/rsyslog.conf`. Default is 'on'
* `node['rackspace_rsyslog']['config']['logs_to_forward']` -  Specifies what logs should be sent to the remote rsyslog server. Default is all ( \*.\* ).
* `node['rackspace_rsyslog']['config']['default_log_dir']` - log directory used in `50-default.conf` template, defaults to `/var/log`
* `node['rackspace_rsyslog']['config']['default_facility_logs']` - Hash containing log facilities and destinations used in `50-default.conf` template.

Recipes
-------
### default
Installs the rsyslog package, manages the rsyslog service and sets up basic configuration for a standalone machine.

### client
Includes `recipe[rackspace_rsyslog]`.

Uses `node['rackspace_rsyslog']['config']['server_ip']` or Chef search (in that precedence order) to determine the remote syslog server's IP address. If search is used, the search query will look for the first `ipaddress` returned from the criteria specified in `node['rackspace_rsyslog']['config']['server_search']`.

If the node itself is a rsyslog server ie it has `rackspace_rsyslog.server` set to true then the configuration is skipped.

If the node had an `/etc/rsyslog.d/35-server-per-host.conf` file previously configured, this file gets removed to prevent duplicate logging.

Any previous logs are not cleaned up from the `log_dir`.

### server
Configures the node to be a rsyslog server. The chosen rsyslog server node should be defined in the `server_ip` attribute or resolvable by the specified search criteria specified in `node['rackspace_rsyslog']['config']['server_search']` (so that nodes making use of the `client` recipe can find the server to log to).

This recipe will create the logs in `node['rackspace_rsyslog']['config']['log_dir']`, and the configuration is in `/etc/rsyslog.d/server.conf`. This recipe also removes any previous configuration to a remote server by removing the `/etc/rsyslog.d/remote.conf` file.

The cron job used in the previous version of this cookbook is removed, but it does not remove any existing cron job from your system (so it doesn't break anything unexpectedly). We recommend setting up logrotate for the logfiles instead.

The `log_dir` will be concatenated with `per_host_dir` to store the logs for each client. Modify the attribute to have a value that is allowed by rsyslogs template matching values, see the rsyslog documentation for this.

Directory structure:

```erb
<%= @log_dir %>/<%= @per_host_dir %>/"logfile"
```

For example for the system with hostname `www`:

```text
/srv/rsyslog/2011/11/19/www/messages
```
For example, to change this to just the hostname, set the attribute `node['rackspace_rsyslog']['config']['per_host_dir']` via a role:

```ruby
"rackspace_rsyslog" => { "config" => { "per_host_dir" => "%HOSTNAME%" } }
```

At this time, the server can only listen on UDP *or* TCP.


Usage
-----
Use `recipe[rackspace_rsyslog]` to install and start rsyslog as a basic configured service for standalone systems.

Use `recipe[rackspace_rsyslog::client]` to have nodes log to a remote server (which is found via the `server_ip` attribute or by the recipe's search call -- see __client__)

Use `recipe[rackspace_rsyslog::server]` to set up a rsyslog server. It will listen on `node['rackspace_rsyslog']['config']['port']` protocol `node['rackspace_rsyslog']['config']['protocol']`.

If you set up a different kind of centralized loghost (syslog-ng, graylog2, logstash, etc), you can still send log messages to it as long as the port and protocol match up with the server software. See __Examples__


### Examples
A `base` role (e.g., roles/base.rb), applied to all nodes so they are syslog clients:

```ruby
name "base"
description "Base role applied to all nodes
run_list("recipe[rackspace_rsyslog::client]")
```

Then, a role for the loghost (should only be one):

```ruby
name "loghost"
description "Central syslog server"
run_list("recipe[rackspace_rsyslog::server]")
```

By default this will set up the clients search for a node with the `loghost` role to talk to the server on TCP port 514. Change the `protocol` and `port` rsyslog attributes to modify this.

If you want to specify another syslog compatible server with a role other than loghost, simply fill free to use the `server_ip` attribute or the `server_search` attribute.

Example role that sets the per host directory:

```ruby
name "loghost"
description "Central syslog server"
run_list("recipe[rackspace_rsyslog::server]")
default_attributes(
  "rackspace_rsyslog" => { "config" => { per_host_dir" => "%HOSTNAME%" } }
)
```

Default rsyslog options are rendered for RHEL family platforms, in `/etc/rsyslog.d/50-default.conf`
with other platforms using a configuration like Debian family defaults.  You can override these
log facilities and destinations using the `rsyslog['default_facility_logs']` hash.

```ruby
name "facility_log_example"
run_list("recipe[rackspace_rsyslog::default]")
default_attributes(
  "rackspace_rsyslog" => {
    "config" => {
      "facility_logs" => {
        '*.info;mail.none;authpriv.none;cron.none' => "/var/log/messages",
        'authpriv' => '/var/log/secure',
        'mail.*' => '-/var/log/maillog',
        '*.emerg' => '*'
      }
    }
  }
)
```

Testing
=======

Please see testing guidelines at [contributing](https://github.com/rackspace-cook
books/contributing/blob/master/CONTRIBUTING.md)

Contributing
============

Please see contributing guidelines at [contributing](https://github.com/rackspac
e-cookbooks/contributing/blob/master/CONTRIBUTING.md)

License & Authors
-----------------
- Author:: Joshua Timberman (<joshua@opscode.com>)
- Author:: Denis Barishev (<denz@twiket.com>)
- Author:: Tim Smith (<tsmith@limelight.com>)
- Author:: Jason Nelson (<jason.nelson@rackspace.com>)

```text
Copyright:: 2009-2013, Opscode, Inc
Copyright:: 2014, Rackspace, US Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
rsyslog Cookbook CHANGELOG
==========================
This file is used to list changes made in each version of the rsyslog cookbook.

v2.0.0
------
- Rewrite for Rackspace specific environments.
- Removed references to OS's not used within Rackspace
- Changed name from rsyslog to rackspace_rsyslog

v1.9.0
------
### New Feature
- **[COOK-3736](https://tickets.opscode.com/browse/COOK-3736)** - Support OmniOS

### Improvement
- **[COOK-3609](https://tickets.opscode.com/browse/COOK-3609)** - Add actionqueue to remote rsyslog configurations

### Bug
- **[COOK-3608](https://tickets.opscode.com/browse/COOK-3608)** - Add 50-default template knobs
- **[COOK-3600](https://tickets.opscode.com/browse/COOK-3600)** - SmartOS support


v1.8.0
------
### Improvement
- **[COOK-3573](https://tickets.opscode.com/browse/COOK-3573)** -  Add Test Kitchen, Specs, and Travis CI

### New Feature
- **[COOK-3435](https://tickets.opscode.com/browse/COOK-3435)** - Add support for relp

v1.7.0
------
### Improvement
- **[COOK-3253](https://tickets.opscode.com/browse/COOK-3253)** - Enable repeated message reduction
- **[COOK-3190](https://tickets.opscode.com/browse/COOK-3190)** - Allow specifying which logs to send to remote server
- **[COOK-2355](https://tickets.opscode.com/browse/COOK-2355)** - Support forwarding events to more than one server

v1.6.0
------
### New Feature
- [COOK-2831]: enable high precision timestamps

### Bug
- [COOK-2377]: calling node.save has adverse affects on nodes relying on a searched node's ohai attributes
- [COOK-2521]: rsyslog cookbook incorrectly sets directory ownership to rsyslog user
- [COOK-2540]: Syslogd needs to be disabled before starting rsyslogd on RHEL 5

### Improvement
- [COOK-2356]: rsyslog service supports status. Service should use it.
- [COOK-2357]: rsyslog cookbook copies in wrong defaults file on Ubuntu !9.10/10.04

v1.5.0
------
- [COOK-2141] - Add `$PreserveFQDN` configuration directive

v1.4.0
------
- [COOK-1877] - RHEL 6 support and refactoring

v1.3.0
------
- [COOK-1189] - template change does not restart rsyslog on Ubuntu

This actually went into 1.2.0 with action `:reload`, but that change has been reverted and the action is back to `:restart`.

v1.2.0
------
- [COOK-1678] - syslog user does not exist on debian 6.0 and ubuntu versions lower than 11.04
- [COOK-1650] - enable max message size configuration via attribute

v1.1.0
------
Changes from COOK-1167:

- More versatile server discovery - use the IP as an attribute, or use search (see README)
- Removed cron dependency.
- Removed log archival; logrotate is recommended.
- Add an attribute to select the per-host directory in the log dir
- Works with Chef Solo now.
- Set debian/ubuntu default user and group. Drop privileges to `syslog.adm`.


v1.0.0
------
- [COOK-836] - use an attribute to specify the role to search for instead of relying on the rsyslog['server'] attribute.
- Clean up attribute usage to use strings instead of symbols.
- Update this README.
- Better handling for chef-solo.


spec/client_spec.rb
require 'spec_helper'

describe 'rsyslog::client' do
  context "when node['rsyslog']['server_ip'] is not set" do
    before do
      Chef::Log.stub(:fatal)
      $stdout.stub(:puts)
    end

    it 'exits fatally' do
      expect { ChefSpec::ChefRunner.new.converge('rsyslog::client') }.to raise_error(SystemExit)
    end
  end

  let(:chef_run) do
    ChefSpec::ChefRunner.new(platform: 'ubuntu', version: '12.04') do |node|
      node.set['rsyslog']['server_ip'] = server_ip
    end.converge('rsyslog::client')
  end

  let(:server_ip) { "10.#{rand(1..9)}.#{rand(1..9)}.50" }
  let(:service_resource) { 'service[rsyslog]' }

  it 'includes the default recipe' do
    expect(chef_run).to include_recipe('rsyslog::default')
  end

  context '/etc/rsyslog.d/49-remote.conf template' do
    let(:template) { chef_run.template('/etc/rsyslog.d/49-remote.conf') }

    it 'creates the template' do
      expect(chef_run).to create_file_with_content(template.path, "*.* @@#{server_ip}:514")
    end

    it 'is owned by root:root' do
      expect(template.owner).to eq('root')
      expect(template.group).to eq('root')
    end

    it 'has 0644 permissions' do
      expect(template.mode).to eq('0644')
    end

    it 'notifies restarting the service' do
      expect(template).to notify(service_resource, :restart)
    end

  end

  context '/etc/rsyslog.d/server.conf file' do
    let(:file) { chef_run.file('/etc/rsyslog.d/server.conf') }

    it 'deletes the file' do
      expect(chef_run).to delete_file(file.path)
    end

    it 'notifies restarting the service' do
      expect(file).to notify(service_resource, :reload)
    end

  end
end


spec/server_spec.rb
require 'spec_helper'

describe 'rsyslog::server' do
  let(:chef_run) do
    ChefSpec::ChefRunner.new(platform: 'ubuntu', version: '12.04') do |node|
      node.set['rsyslog']['server'] = false
    end.converge('rsyslog::server')
  end

  let(:service_resource) { 'service[rsyslog]' }

  it "sets node['rsyslog']['server'] to true" do
    expect(chef_run.node['rsyslog']['server']).to be_true
  end

  it 'includes the default recipe' do
    expect(chef_run).to include_recipe('rsyslog::default')
  end

  context '/srv/rsyslog directory' do
    let(:directory) { chef_run.directory('/srv/rsyslog') }

    it 'creates the directory' do
      expect(chef_run).to create_directory('/srv/rsyslog')
    end

    it 'is owned by root:root' do
      expect(directory.owner).to eq('root')
      expect(directory.group).to eq('root')
    end

    it 'has 0755 permissions' do
      expect(directory.mode).to eq('0755')
    end
  end

  context '/etc/rsyslog.d/35-server-per-host.conf template' do
    let(:template) { chef_run.template('/etc/rsyslog.d/35-server-per-host.conf') }

    it 'creates the template' do
      expect(chef_run).to create_file_with_content(template.path, '/srv/rsyslog/%$YEAR%/%$MONTH%/%$DAY%/%HOSTNAME%/auth.log')
    end

    it 'is owned by root:root' do
      expect(template.owner).to eq('root')
      expect(template.group).to eq('root')
    end

    it 'has 0644 permissions' do
      expect(template.mode).to eq('0644')
    end

    it 'notifies restarting the service' do
      expect(template).to notify(service_resource, :restart)
    end

  end

  context '/etc/rsyslog.d/remote.conf file' do
    let(:file) { chef_run.file('/etc/rsyslog.d/remote.conf') }

    it 'deletes the file' do
      expect(chef_run).to delete_file(file.path)
    end

    it 'notifies restarting the service' do
      expect(file).to notify(service_resource, :reload)
    end

  end
end


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'
require 'chef/application'

at_exit { ChefSpec::Coverage.report! }


recipes/server.rb
#
# Cookbook Name:: rackspace_rsyslog
# Recipe:: server
#
# Copyright 2009-2013, Opscode, Inc.
# Copyright 2014, Rackspace, US Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Manually set this attribute
node.set['rackspace_rsyslog']['config']['server'] = true

include_recipe 'rackspace_rsyslog::default'

directory node['rackspace_rsyslog']['config']['log_dir'] do
  owner    'root'
  group    'root'
  mode     '0755'
  recursive true
end

template "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d/35-server-per-host.conf" do
  cookbook node['rackspace_rsyslog']['templates']['35-server-per-host.conf']
  source   '35-server-per-host.conf.erb'
  owner    'root'
  group    'root'
  mode     '0644'
  notifies :restart, "service[#{node['rackspace_rsyslog']['service_name']}]"
end

file "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d/remote.conf" do
  action   :delete
  notifies :reload, "service[#{node['rackspace_rsyslog']['service_name']}]"
  only_if  { ::File.exists?("#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d/remote.conf") }
end


recipes/default.rb
#
# Cookbook Name:: rackspace_rsyslog
# Recipe:: default
#
# Copyright 2009-2013, Opscode, Inc.
# Copyright 2014, Rackspace, US Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

package 'rsyslog'
package 'rsyslog-relp' if node['rackspace_rsyslog']['config']['use_relp']

directory "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d" do
  owner 'root'
  group 'root'
  mode  '0755'
end

directory '/var/spool/rsyslog' do
  owner 'root'
  group 'root'
  mode  '0755'
end

# Our main stub which then does its own rsyslog-specific
# include of things in /etc/rsyslog.d/*
template "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.conf" do
  cookbook node['rackspace_rsyslog']['templates']['rsyslog.conf']
  source  'rsyslog.conf.erb'
  owner   'root'
  group   'root'
  mode    '0644'
  notifies :restart, "service[#{node['rackspace_rsyslog']['service_name']}]"
end

template "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d/50-default.conf" do
  cookbook node['rackspace_rsyslog']['templates']['50-default.conf']
  source  '50-default.conf.erb'
  owner   'root'
  group   'root'
  mode    '0644'
  notifies :restart, "service[#{node['rackspace_rsyslog']['service_name']}]"
end

service node['rackspace_rsyslog']['service_name'] do
  supports restart: true, reload: true, status: true
  action   [:enable, :start]
end


recipes/client.rb
#
# Cookbook Name:: rackspace_rsyslog
# Recipe:: client
#
# Copyright 2009-2013, Opscode, Inc.
# Copyright 2014, Rackspace, US Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Do not run this recipe if the server attribute is set
return if node['rackspace_rsyslog']['config']['server']

include_recipe 'rackspace_rsyslog::default'

# On Chef Solo, we use the node['rackspace_rsyslog']['config'][:server_ip] attribute, and on
# normal Chef, we leverage the search query.
if Chef::Config[:solo]
  if node['rackspace_rsyslog']['config']['server_ip']
    rsyslog_servers = Array(node['rackspace_rsyslog']['config']['server_ip'])
  else
    Chef::Application.fatal!("Chef Solo does not support search. You must set node['rackspace_rsyslog']['config']['server_ip']!")
  end
else
  results = search(:node, node['rackspace_rsyslog']['config']['server_search']).map { |n| n[:ipaddress] }
  rsyslog_servers = Array(node['rackspace_rsyslog']['config']['server_ip']) + Array(results)
end

if rsyslog_servers.empty?
  Chef::Application.fatal!('The rsyslog::client recipe was unable to determine the remote syslog server. Checked both the server_ip attribute and search!')
end

remote_type = node['rackspace_rsyslog']['config']['use_relp'] ? 'relp' : 'remote'

template "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d/49-remote.conf" do
  cookbook  node['rackspace_rsyslog']['templates']['49-#{remote_type}.conf']
  source    "49-#{remote_type}.conf.erb"
  owner     'root'
  group     'root'
  mode      '0644'
  variables(servers: rsyslog_servers)
  notifies  :restart, "service[#{node['rackspace_rsyslog']['service_name']}]"
  only_if   { node['rackspace_rsyslog']['config']['remote_logs'] }
end

file "#{node['rackspace_rsyslog']['config']['config_prefix']}/rsyslog.d/server.conf" do
  action   :delete
  notifies :reload, "service[#{node['rackspace_rsyslog']['service_name']}]"
end
